File: backend/app/Models/School.php
<?php

namespace App\Models;

use App\Enums\SchoolStatus;
use App\Support\SchoolClock;
use Database\Factories\SchoolFactory;
use Illuminate\Database\Eloquent\Attributes\Fillable;
use Illuminate\Database\Eloquent\Factories\HasFactory;
use Illuminate\Database\Eloquent\Model;
use Illuminate\Database\Eloquent\Relations\BelongsTo;
use Illuminate\Database\Eloquent\Relations\HasMany;
use Illuminate\Database\Eloquent\Relations\HasOne;

#[Fillable([
    'parent_school_id',
    'name', 'registration_number', 'email', 'phone', 'address', 'city',
    'state', 'country', 'postal_code', 'latitude', 'longitude',
    'currency_code', 'timezone', 'logo_url',
    'status',
])]
// 'status' is fillable here for the same reason as User::status (Phase 1):
// SchoolService sets it explicitly on create/activate/deactivate, and the
// general update() request never includes it in its validated payload.
class School extends Model
{
    /** @use HasFactory<SchoolFactory> */
    use HasFactory;

    /**
     * Memoised by groupSchoolIds(). Not an attribute - it is derived, and
     * lives only as long as this instance does.
     *
     * @var array<int, int>|null
     */
    private ?array $groupSchoolIds = null;

    protected function casts(): array
    {
        return [
            'status' => SchoolStatus::class,
            'latitude' => 'decimal:7',
            'longitude' => 'decimal:7',
        ];
    }

    /**
     * The group this school belongs to, if it is a branch of one.
     *
     * @return BelongsTo<School, $this>
     */
    public function parent(): BelongsTo
    {
        return $this->belongsTo(School::class, 'parent_school_id');
    }

    /**
     * The branches beneath this school, if it is a parent.
     *
     * @return HasMany<School, $this>
     */
    public function branches(): HasMany
    {
        return $this->hasMany(School::class, 'parent_school_id');
    }

    public function isBranch(): bool
    {
        return $this->parent_school_id !== null;
    }

    /**
     * Every school in this one's group - itself, plus its branches if it is a
     * parent, or its parent and siblings if it is a branch.
     *
     * Deliberately one level deep: a group of groups would need a recursive
     * query here and everywhere that calls it, and buys nothing a school has
     * asked for. Validation stops the data ever becoming deeper than this.
     *
     * Held on the instance after the first call. Every School Admin's scope
     * now resolves through here, and a scope is built once per policy check -
     * so a list of fifty students would otherwise ask the same question fifty
     * times (CLAUDE.md rule 22).
     *
     * @return array<int, int>
     */
    public function groupSchoolIds(): array
    {
        $rootId = $this->parent_school_id ?? $this->id;

        return $this->groupSchoolIds ??= School::query()
            ->where('id', $rootId)
            ->orWhere('parent_school_id', $rootId)
            ->orderBy('id')
            ->pluck('id')
            ->all();
    }

    /**
     * @return HasMany<User, $this>
     */
    public function users(): HasMany
    {
        return $this->hasMany(User::class);
    }

    /**
     * @return HasOne<CommunicationSetting, $this>
     */
    public function communicationSetting(): HasOne
    {
        return $this->hasOne(CommunicationSetting::class);
    }

    /**
     * What time it is at this school. Anything that decides a date - "is this
     * today?", "what month is this?" - asks here rather than the server.
     */
    public function clock(): SchoolClock
    {
        return SchoolClock::for($this);
    }

    public function isActive(): bool
    {
        return $this->status === SchoolStatus::Active;
    }
}

File: frontend/lib/features/schools/data/models/school.dart
enum SchoolStatus {
  active('active'),
  inactive('inactive');

  const SchoolStatus(this.apiValue);

  final String apiValue;

  static SchoolStatus fromApiValue(String value) => SchoolStatus.values.firstWhere((s) => s.apiValue == value);
}

class School {
  const School({
    required this.id,
    required this.name,
    this.parentSchoolId,
    this.parentSchoolName,
    this.branchCount = 0,
    required this.registrationNumber,
    required this.email,
    required this.phone,
    required this.address,
    required this.city,
    required this.state,
    required this.country,
    required this.postalCode,
    this.latitude,
    this.longitude,
    required this.currencyCode,
    required this.timezone,
    required this.logoUrl,
    required this.status,
  });

  factory School.fromJson(Map<String, dynamic> json) {
    return School(
      id: json['id'] as int,
      name: json['name'] as String,
      parentSchoolId: json['parent_school_id'] as int?,
      parentSchoolName: json['parent_school_name'] as String?,
      branchCount: json['branch_count'] as int? ?? 0,
      registrationNumber: json['registration_number'] as String?,
      email: json['email'] as String,
      phone: json['phone'] as String,
      address: json['address'] as String,
      city: json['city'] as String,
      state: json['state'] as String,
      country: json['country'] as String,
      postalCode: json['postal_code'] as String,
      latitude: json['latitude'] as String?,
      longitude: json['longitude'] as String?,
      currencyCode: json['currency_code'] as String,
      timezone: json['timezone'] as String? ?? 'UTC',
      logoUrl: json['logo_url'] as String?,
      status: SchoolStatus.fromApiValue(json['status'] as String),
    );
  }

  final int id;
  final String name;

  /// The school this one is a branch of, if any. See docs/branches.md.
  final int? parentSchoolId;
  final String? parentSchoolName;

  /// How many branches sit beneath this one. Zero for a branch, and for a
  /// standalone school.
  final int branchCount;

  bool get isBranch => parentSchoolId != null;

  bool get isGroupParent => branchCount > 0;
  final String? registrationNumber;
  final String email;
  final String phone;
  final String address;
  final String city;
  final String state;
  final String country;
  final String postalCode;

  /// Where the school is, if anyone has said. Kept as the string the
  /// API sent so a coordinate is never quietly rounded by a double on
  /// its way through the client.
  final String? latitude;
  final String? longitude;
  final String currencyCode;

  /// The IANA zone this school's dates and times are read in, e.g.
  /// `Asia/Kolkata`. Decides what "today" means for everything it records.
  final String timezone;
  final String? logoUrl;
  final SchoolStatus status;

  School copyWith({SchoolStatus? status}) {
    return School(
      id: id,
      name: name,
      registrationNumber: registrationNumber,
      email: email,
      phone: phone,
      address: address,
      city: city,
      state: state,
      country: country,
      postalCode: postalCode,
      latitude: latitude,
      longitude: longitude,
      currencyCode: currencyCode,
      timezone: timezone,
      logoUrl: logoUrl,
      status: status ?? this.status,
    );
  }
}
